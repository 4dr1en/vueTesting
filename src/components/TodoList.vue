<script setup lang="ts">
import { ref } from "vue";
import TodoForm from "./TodoForm.vue";
import TodoEdit from "./TodoEdit.vue";

const todos = ref<string[]>([]);
const currentEdit = ref<number | null>(null);

function removeTodo(index: number) {
  todos.value.splice(index, 1);
}

function editTodo(index: number, text: string) {
  currentEdit.value = null;
  todos.value[index] = text;
}
</script>

<template>
  <div>
    <h1>LIST OF TODO</h1>
    <ul class="todo-list">
      <li v-for="(todo, index) in todos">
        <p>{{ todo }}</p>
        <button @click="removeTodo(index)" class="todo__delete">Delete</button>
        <button
          @click="
            currentEdit == index ? (currentEdit = null) : (currentEdit = index)
          "
          class="todo__edit"
        >
          {{ currentEdit == index ? "Cancel edit" : "Edit" }}
        </button>
        <TodoEdit
          v-if="index == currentEdit"
          :oldTodo="todo"
          @editTodo="editTodo(index, $event)"
        />
      </li>
    </ul>
    <TodoForm @addTodo="(e: string) => todos.push(e)" />
  </div>
</template>

<style lang="scss"></style>
