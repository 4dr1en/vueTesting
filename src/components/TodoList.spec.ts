import { mount } from "@vue/test-utils";
import TodoList from "@/components/TodoList.vue";
import { expect, describe, it } from "vitest";

describe("TodoList", () => {
  it("mount an empty list", () => {
    const wrapper = mount(TodoList);
    const todoItems = wrapper.findAll("li");
    expect(todoItems).toHaveLength(0);
  });

  it("adds a new todo when the form is submitted", async () => {
    const wrapper = mount(TodoList);

    const todoForm = wrapper.findComponent({ name: "TodoForm" });
    const input = todoForm.find("input");
    const newTodo = "Buy groceries";

    await input.setValue(newTodo);
    await todoForm.find("button").trigger("click");

    expect(wrapper.find(".todo-list").text()).toContain(newTodo);
  });

  it("removes a todo when the remove button is clicked", async () => {
    const wrapper = mount(TodoList);

    const todoForm = wrapper.findComponent({ name: "TodoForm" });
    const input = todoForm.find("input");
    const newTodo = "Buy groceries";

    await input.setValue(newTodo);
    await wrapper.find("button").trigger("click");

    const removeButton = wrapper.find("button.todo__delete");
    await removeButton.trigger("click");

    expect(wrapper.find(".todo-list").text()).not.toContain(newTodo);
  });

  it("Display the 'TodoEdit' component when the edit button is clicked", async () => {
    const wrapper = mount(TodoList);

    const todoForm = wrapper.findComponent({ name: "TodoForm" });
    const input = todoForm.find("input");
    const newTodo = "Buy groceries";
    await input.setValue(newTodo);
    await wrapper.find("button").trigger("click");

    const editButton = wrapper.find("button.todo__edit");
    await editButton.trigger("click");

    expect(wrapper.findComponent({ name: "TodoEdit" }).exists()).toBe(true);
  });

  it("Should match the snapshot", () => {
    const wrapper = mount(TodoList);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
