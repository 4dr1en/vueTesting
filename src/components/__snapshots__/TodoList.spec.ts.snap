// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList > Should match the snapshot 1`] = `
"<div>
  <h1>LIST OF TODO</h1>
  <ul class=\\"todo-list\\"></ul>
  <div><label for=\\"todo\\">Todo</label><input type=\\"text\\" id=\\"todo\\"><button>Add</button></div>
</div>"
`;
